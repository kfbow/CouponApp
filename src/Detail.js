/** @flow */
'use strict';

import React from 'react';
import {
    Alert,
    Clipboard,
    View,
    Text,
    Image,
    Linking,
    ScrollView,
    StatusBar,
    TouchableOpacity
} from 'react-native';
import Dimensions from 'Dimensions';
import { stringDiscounted, stringOriginal } from './RandomStrings';

export const dimensions = Dimensions.get('window');

const styles = {
    categoryContainer: {
        borderRadius: 20,
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: '#CEEAE7',
        backgroundColor: '#CEEAE7',
        paddingTop: 4,
        paddingBottom: 4,
        paddingLeft: 8,
        paddingRight: 8,
    },
    container: {
        padding: 10,
    },
    description: {
        marginTop: 15,
    },
    expire: {
        alignSelf: 'center',
        color: '#ccc'
    },
    image: {
        width: dimensions.width * (9/10),
        height: dimensions.width * (9/10),
        resizeMode: 'contain',
        alignSelf: 'center'
    },
    name: {
        fontSize: 30,
        marginTop: 20,
    },
    greyOut: {
        color: '#ccc',
    },
    priceAmazon: {
        flexDirection: 'row',
        marginTop: 20,
        justifyContent: 'space-around'
    },
    promoContainer: {
        position: 'absolute',
        backgroundColor: '#fff',
        bottom: 0,
        paddingBottom: 10,
        paddingTop: 10
    },
    promoPrice: {
        fontSize: 20,
    },
    promoReveal: {
        padding: dimensions.height * (.2/10),
        textAlign: 'center',
        backgroundColor: '#159588',
        color: '#fff',
        width: dimensions.width * .9,
        marginRight: dimensions.width * .05,
        marginLeft: dimensions.width * .05,
    },
    seeAmazon: {
        padding: 15,
        backgroundColor: '#558EFC',
        color: '#fff'
    },
    separator: {
        height: 1,
        backgroundColor: '#d3d3d3',
        marginTop: 5,
    }
}

/*
// Open an alert giving the user the option to go to Amazon or Cancel while
// setting the promo code to their clipboard
*/
const onButtonPress = (data) => {
    Alert.alert(
        `Copied code ${data.promoCode}`,
        'Promo code copied to your clipboard',
        [
            {text: 'Cancel', onPress: () => console.log('Cancel Pressed!')},
            {text: 'Go To Amazon',
                onPress: () => openAmazon(data)
            },
        ]
    );
    Clipboard.setString(data.promoCode);
}

// Open the Amazon app or webpage in Safari if app is not installed
const openAmazon = (data) => {
    return Linking
        .openURL(data.externalUrl)
        .catch(err => {
            console.log('An error occurred', err);
        })
}

const onClose = (setStateObject, data) => {
    setStateObject('showModal', false, data);
}

const Detail = (props) => {
    const { showModal, setStateObject, data } = props;

    return (
        <View style={{ flex: 1 }}>
            <StatusBar
                hidden={ true } />
            <ScrollView
                style={ styles.container }
                contentInset={{ bottom: dimensions.height * .1 }}>
                <Text
                    style={{ padding: 20 }}
                    onPress={ () => onClose(setStateObject, data) }>
                    Close
                </Text>
                <Image
                    source={{uri: `https:${data.imageUrl}`}}
                    style={ styles.image }/>

                <View style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
                    <View style={ styles.categoryContainer }>
                        <Text style={ styles.category }>
                            { data.categoryName }
                        </Text>
                    </View>
                    <Text style={ styles.expire }>
                        { data.expiresAt }
                    </Text>
                </View>

                <View style={ styles.separator } />

                <View>
                    <Text style={ styles.name }>
                        { data.name }
                    </Text>
                    <Text style={ styles.description }>
                        { data.description }
                    </Text>
                    <View style={ styles.priceAmazon }>
                        <View>
                            <Text style={ styles.greyOut }>
                                {`${stringOriginal()} `}
                                <Text style={{textDecorationLine: 'line-through'}}>
                                    ${ data.price }
                                </Text>
                            </Text>
                            <Text style={ styles.promoPrice }>
                                {stringDiscounted()} ${data.promoPrice}
                            </Text>
                        </View>
                        <Text
                            style={ styles.seeAmazon }
                            onPress={ () => openAmazon(data) }>
                            See on Amazon
                        </Text>
                    </View>
                </View>
            </ScrollView>

            <TouchableOpacity
                onPress={ () => onButtonPress(data) }>
                <View style={ styles.promoContainer }>
                    <Text style={ styles.promoReveal }>
                        Reveal Promo Code
                    </Text>
                </View>
            </TouchableOpacity>

        </View>
    )
}

export default Detail;
