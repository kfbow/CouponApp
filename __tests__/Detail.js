/**
 * Jest tests for the Detail component
 */

'use strict';

import 'react-native';
import React from 'react';
import Detail from '../src/Detail';
import data from './response.json';

import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <Detail data={ data }/>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders initial price correctly', () => {
  const Text = require('Text');
  const tree = renderer.create(
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          {
            "textDecorationLine": "line-through",
          }
        }>
        $
        29.00
      </Text>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders promo price correctly', () => {
  const Text = require('Text');
  const tree = renderer.create(
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          {
            "fontSize": 20,
          }
        }>
        only here
         $
        4.00
      </Text>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders item category correctly', () => {
  const Text = require('Text');
  const tree = renderer.create(
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={undefined}>
        Beauty & Health
      </Text>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
