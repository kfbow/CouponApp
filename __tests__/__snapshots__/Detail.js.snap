exports[`test renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }>
  <ScrollView
    style={
      Object {
        "padding": 10,
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "padding": 20,
        }
      }>
      Close
    </Text>
    <Image
      source={
        Object {
          "uri": "https://snagshoutprod-media.s3.amazonaws.com/e837f68e8db72df44cc2c2a0ce47e98c8b7778e2.jpg",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "height": 337.5,
          "resizeMode": "contain",
          "width": 337.5,
        }
      } />
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }>
      <View
        style={
          Object {
            "backgroundColor": "#CEEAE7",
            "borderColor": "#CEEAE7",
            "borderRadius": 20,
            "borderStyle": "solid",
            "borderWidth": 1,
            "paddingBottom": 4,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 4,
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail">
          Beauty & Health
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#ccc",
          }
        }>
        November 30, 2016
      </Text>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#d3d3d3",
          "height": 1,
          "marginTop": 5,
        }
      } />
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 30,
            "marginTop": 20,
          }
        }>
        NuPeptin Lengthening Eye Lash Serum - stimulates l...
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "marginTop": 15,
          }
        }>
        NuPeptin Eye Lash Serum helps stimulate lash growth. For visibly thicker and longer eyelashes. Our Eye Lash Serum is made without parabens and is hypoallergenic. Ophthalmologists recommend this product for lengthening thinning eyelashes and dermatologists have tested it. Our clinical studies have show that after 2 weeks you can expect a 4mm lengthening of the eyelash and in 4 weeks an 8mm lengthening. When testing the active ingredient alone, there was a 25% increase in length in just 2 weeks. For best results, remove makeup with a mild cleaner. At bedtime, apply product as you would liquid eyeliner at the base of the upper and lower lash line. The amount of liquid on the brush is enough for the upper and lower lash line of both eyes.
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-around",
            "marginTop": 20,
          }
        }>
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ccc",
              }
            }>
            regular price 
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "textDecorationLine": "line-through",
                }
              }>
              $
              29.00
            </Text>
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 20,
              }
            }>
            with coupon
             $
            4.00
          </Text>
        </View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          onPress={[Function]}
          style={
            Object {
              "backgroundColor": "#558EFC",
              "color": "#fff",
              "padding": 15,
            }
          }>
          See on Amazon
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "marginBottom": 66.7,
        }
      } />
  </ScrollView>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "bottom": 0,
        "paddingBottom": 10,
        "paddingTop": 10,
        "position": "absolute",
      }
    }>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "backgroundColor": "#159588",
            "color": "#fff",
            "marginLeft": 18.75,
            "marginRight": 18.75,
            "padding": 13.34,
            "textAlign": "center",
            "width": 337.5,
          }
        }>
        Reveal Promo Code
      </Text>
    </View>
  </View>
</View>
`;

exports[`test renders initial price correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "textDecorationLine": "line-through",
    }
  }>
  $ 29.00
</Text>
`;

exports[`test renders item category correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={undefined}>
  Beauty & Health
</Text>
`;

exports[`test renders promo price correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "fontSize": 20,
    }
  }>
  only here $ 4.00
</Text>
`;
